Skip non-update endpoint updates

joelsmith opened the pull request:
**What this PR does / why we need it**:

On large clusters, a large percentage of endpoint updates are actually non-updates that occur as a result of a change in an associated pod. This results in endpoint updates where the only field that has changed is the `TargetRef.ResourceVersion` in the endpoint address associated with the changed pod. Given enough of these non-updates, the endpoint controller's queue rate limit can be overwhelmed and legitimate updates can be delayed, resulting in (temporarily) broken services. We have clusters where we've seen endpoint updates take 9 minutes.

**Which issue this PR fixes** : fixes #50936 

**Special notes for your reviewer**:
N/A

**Release note**:
```release-note
Prevent unneeded endpoint updates
```

sjenning committed:
Factor out endpoint address generation, skip unneeded endpoint updates

Also add unit tests for new endpoint helpers related to updatePod

joelsmith commented:
@sjenning @DirectXMan12 PTAL. Also, could somebody please add the ok to test label?

sjenning commented:
/ok-to-test

sjenning commented:
Darn it Joel! You made me pull out pen and paper for a Venn diagram :stuck_out_tongue:

This looks functionally correct to me, but it is hard to follow in places and does some things unnecessarily.  For example, if we know that pod and labels didn't change, we can immediately return as we know there is no update required, avoiding one or both `getPodServiceMemberships()` calls which are pretty expensive.

The changes were harder to describe in words than in code, so just adpated your code (untested):
https://github.com/sjenning/kubernetes/commit/7bcbe3811eb575ec468806031a00a69023fb8845

I think that does the minimal amount of work required and might be easier to understand.  What do you think?

also cc @derekwaynecarr @eparis @smarterclayton 

Looks like all the tests are passing except for one flake so that's nice.

Great work running this one down! :+1:

joelsmith commented:
@sjenning I like your updated version. I'll test it out then pull it in, squash and push to this PR's branch. Hopefully the flake will work the second time. Thanks for your help!

thockin commented on pkg/controller/endpoint/endpoints_controller.go:
```diff
@@ -213,22 +226,57 @@ func (e *EndpointController) updatePod(old, cur interface{}) {
+	newEndpointAddress.NodeName = nil
+	oldEndpointAddress.NodeName = nil
+	if !reflect.DeepEqual(newEndpointAddress, oldEndpointAddress) {
+		// The pod has not changed in any way that impacts the endpoints
```
s/has not/has  ??

joelsmith replied:
Good catch, we let an outdated comment slip through

thockin reviewed (commented):
This looks great overall, but it really needs a test.  At the very least, we can break out the comparison logic to a function and test against various inputs, and we can break out the service-change logic into a function, and test against various scenarios to prove that your set logic stays correct.

joelsmith commented:
@thockin Thanks for your review. I have added the unit tests you recommended. I was planning on adding some tests (the reason for the "WIP" in the title) but I was struggling to come up with anything good. I appreciate your suggestions and I hope the newly-added tests are adequate.

thockin commented on pkg/controller/endpoint/endpoints_controller.go:
```diff
@@ -202,6 +202,48 @@ func (e *EndpointController) addPod(obj interface{}) {
+	oldEndpointAddress := podToEndpointAddress(oldPod)
+	newEndpointAddress.TargetRef.ResourceVersion = ""
+	oldEndpointAddress.TargetRef.ResourceVersion = ""
+	newEndpointAddress.NodeName = nil
```
Comment why we don't care about NodeName but we do care about TargetRef?

thockin commented on pkg/controller/endpoint/endpoints_controller_test.go:
```diff
@@ -956,3 +957,88 @@ func TestShouldPodBeInEndpoints(t *testing.T) {
+		t.Errorf("Expected address to be unchanged for copied pod")
+	}
+
+	newPod.ObjectMeta.ResourceVersion = "changed"
```
repeat this for NodeName, if we really don't care about that (not convinced we don't).

thockin commented on pkg/controller/endpoint/endpoints_controller_test.go:
```diff
@@ -956,3 +957,88 @@ func TestShouldPodBeInEndpoints(t *testing.T) {
+	bcd := sets.NewString("b", "c", "d")
+	abcd := sets.NewString("a", "b", "c", "d")
+	ad := sets.NewString("a", "d")
+	retval := determineNeededServiceUpdates(abc, bcd, false)
```
Can you do this as a table-driven test and exercise more cases.  One or the other being empty.  Both empty.  Totally disjoint sets, identical sets, etc.

Let me know if you need examples of such tests.

thockin reviewed (commented):
thanks.  Almost there.  Any time there is something non-obvious, err on the side of more comments :)

joelsmith commented:
@sjenning Were you ignoring the `NodeName` because of what I said about its value being different? If so, I think I led us down a bad path. My earlier iteration of an equality function to use in place of `DeepEqual` didn't handle NodeName, but `DeepEqual` does, and in my testing, it doesn't cause unnecessary endpoint updates. It looks like that was all due to the `TargetRef.ResourceVersion`. Sorry for the confusion. I've made the latest version stop ignoring `NodeName` and everything I've tested is working. Please let me know if there is something we're missing w.r.t. `NodeName`, but I suspect that you just added it to the ignore list due to my comments about my earlier mistaken implementation.

@thockin I've switched to the table of test cases and added a few more test cases. Please let me know if you think of any others that I missed.

thockin commented:
This LGTM.  Can you please squash commits?

joelsmith commented:
@thockin squashed, and thanks again for the review!

thockin commented:
/lgtm
/approve

joelsmith commented:
Sorry for the extra churn. I should have been more careful while editing code on a test VM. In the latest squashed version I've run gofmt against `endpoints_controller.go`. That's the only difference and hopefully we'll pass all the automated tests this time!

joelsmith commented:
/test pull-kubernetes-kubemark-e2e-gce

eparis commented:
/retest
/lgtm

@joelsmith @thockin @derekwaynecarr @sjenning 
This is obviously a large improvement. But I'm not convinced it is the "right" fix. I question the value of the pod's ResourceVersion in the endpoint object at all. What is it buying us? It is the main reason we have this patch right? And is a rather large write load on etcd.

Maybe we can't get rid of it because it is definitely in the AP. But, can/should we start a conversation about just never setting `EndpointAddress.TargetRef.ResourceVersion` at all? Make it ALWAYS just "" ? Who do we think could approve killing part of the API? Or who could tell us why ResourceVersion should be there at all?

sjenning commented:
pull-kubernetes-kubemark-e2e-gce is completely busted at the moment.  All PRs are failing and the submit queue is blocked.  No PR has passed this test since 2pm CDT on 8/19.

sjenning commented:
looks like kubemark-e2e-gce might be back online
/retest

derekwaynecarr commented:
@eparis -- i believe this will update the resource version whenever a meaningful change was made, but I agree from my perspective in the long run the resource version of the pod is not super meaningful, and at worse makes people want to compare them.

@joelsmith - change lgtm as well, thx for diving into this problem.
